api/ccache_test.go

```diff
@@ -0,0 +1,203 @@
+package api
+
+import (
+	"bytes"
+	"encoding/json"
+	"net/http"
+	"net/http/httptest"
+	"testing"
+	"time"
+
+	"github.com/grafana/metrictank/api/models"
+	"github.com/grafana/metrictank/api/response"
+	"github.com/grafana/metrictank/cluster"
+	"github.com/grafana/metrictank/conf"
+	"github.com/grafana/metrictank/idx/memory"
+	"github.com/grafana/metrictank/mdata"
+	"github.com/grafana/metrictank/mdata/cache"
+	"gopkg.in/raintank/schema.v1"
+)
+
+func newSrv(delSeries, delArchives int, key string) (*Server, *cache.MockCache) {
+	srv, _ := NewServer()
+	srv.RegisterRoutes()
+
+	mdata.SetSingleAgg(conf.Avg, conf.Min, conf.Max)
+	mdata.SetSingleSchema(conf.NewRetentionMT(10, 100, 600, 10, true))
+
+	store := mdata.NewDevnullStore()
+	srv.BindBackendStore(store)
+
+	mockCache := cache.NewMockCache()
+	mockCache.DelMetricSeries = delSeries
+	mockCache.DelMetricArchives = delArchives
+	metrics := mdata.NewAggMetrics(store, mockCache, false, 0, 0, 0)
+	srv.BindMemoryStore(metrics)
+	srv.BindCache(mockCache)
+
+	metricIndex := memory.New()
+	metricIndex.AddOrUpdate(
+		&schema.MetricData{
+			Id:       key,
+			OrgId:    1,
+			Name:     "test.key",
+			Metric:   "test.key",
+			Interval: 10,
+			Value:    1,
+		},
+		0,
+	)
+	srv.BindMetricIndex(metricIndex)
+	return srv, mockCache
+}
+
+func TestMetricDelete(t *testing.T) {
+	cluster.Init("default", "test", time.Now(), "http", 6060)
+
+	delSeries := 3
+	delArchives := 10
+	testKey := "12345"
+
+	srv, cache := newSrv(delSeries, delArchives, testKey)
+	req, _ := json.Marshal(models.CCacheDelete{
+		Patterns:  []string{"test.*"},
+		OrgId:     1,
+		Propagate: false,
+	})
+
+	ts := httptest.NewServer(srv.Macaron)
+	defer ts.Close()
+
+	res, err := http.Post(ts.URL+"/ccache/delete", "application/json", bytes.NewReader(req))
+	if err != nil {
+		t.Fatalf("There was an error in the request: %s", err)
+	}
+
+	respParsed := models.CCacheDeleteResp{}
+	buf := new(bytes.Buffer)
+	buf.ReadFrom(res.Body)
+	json.Unmarshal(buf.Bytes(), &respParsed)
+
+	if len(cache.DelMetricKeys) != 1 || cache.DelMetricKeys[0] != testKey {
+		t.Fatalf("Expected that key %s has been deleted, but it has not", testKey)
+	}
+
+	if respParsed.DeletedSeries != delSeries || respParsed.DeletedArchives != delArchives {
+		t.Fatalf("Expected %d series and %d archives to get deleted, but got %d and %d", delSeries, delArchives, respParsed.DeletedSeries, respParsed.DeletedArchives)
+	}
```

Comment:
i'm confused about why we tell the cache to return these fake, made up numbers and then check them, while we also check the real numbers above. can this be simplified ? for example have the mockcache return that deleted series == deleted metric keys, and deleted archives is maybe pinned to 3x the number of deleted keys? does this idea make sense? if not, add a comment somewhere that explains this please

Reply:
This is only testing the request handler, not the cache. It's testing one request with propagation disabled (`TestMetricDelete`) and one with propagation enabled (`TestMetricDeletePropagation`).  I'm not sure how creating additional rules like `archives = 3 * series` would simplify things, that would rather just make it more complicated because in order to understand the test a reader would then first need to be aware of this rule.
Maybe I should rename the tests to `TestMetricDeleteRequestHandlerWithoutPropagation()` and `TestMetricDeleteRequestHandlerWithPropagation`?

Reply:
it's confusing because the mock cache claims to have deleted a number series and archives that has nothing to do with the delete request we actually issued on it. the numbers don't seem to make sense. we assert that it only received 1 metric key delete , then how could any cache have deleted 3 series if there was only 1 key? i know it's a mock and not a real cache, but the numbers should make more sense I think.

Reply:
ah, ok i understand what you mean

Reply:
how's that https://github.com/grafana/metrictank/pull/555/commits/58d279a619d62cdac74b1f286436f19d0356d693

Reply:
that would clear things up if the DelMetric method took patterns.
but according to the argument names or the interface function and its implementations
(and the docs for CCache.DelMetric) it only takes 1 metric key, not a pattern.

Reply:
i found a bug while looking at that just now: https://github.com/grafana/metrictank/pull/555/commits/b168558e98f56d0dc7d4aaca2e99772cd6c7b3b4